engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/id/execution/IdExecutionInstrument.java

```diff
@@ -180,15 +203,18 @@ private class IdExecutionEventNode extends ExecutionEventNode {
 
       private final EventContext context;
       private long nanoTimeElapsed = 0;
-      private ExecutionEnvironment originalExecutionEnvironment = null;
+      private ExecutionEnvironment originalExecutionEnvironment;
 
       /**
-       * Creates a new event node.
+       * Creates a new event node for instrumentation.
        *
        * @param context location where the node is being inserted
+       * @param inheritedEnvironment if non-{@code null} then value carries over environment
+       *     information from the previously invalidated event node for the same underlying node
        */
-      IdExecutionEventNode(EventContext context) {
+      IdExecutionEventNode(EventContext context, ExecutionEnvironment inheritedEnvironment) {
```

Comment:
- The special handling of `ExecutionEnvironment` is a relict
- #12340 suggests to change it
- I am afraid this `inheritedEnvironment` isn't getting us closer to #12340 vision